### DISCUSSION TITLE
Can getStaticProps be thought of as an alternative to useState hook to fetch data before page load

### DISCUSSION THREAD

**Discussion by Jordan-Rob** [vercel/next.js#25157] [Help]
Hi everyone, I'm new to next js and I'm currently going through the official tutorial on nextjs.org. I'm on the implement getStaticProps part under pre-rendering and Data fetching, and according to what I have read, I'm wondering whether getStaticProps can come in as an alternative to using the React hook useState to fetch data before page load.

**Comment by JacobLey** [comment]
By `useState` you mean something like:
```
const [myData, setMyData] = useState(null);
useEffect(async () => {
   const loadedData = await fetchLoadedData();
   setMyData(loadedData)
});
```
?

This pattern is useful because you have to draw something right away, and will show data once loaded.

If the case is `fetchLoadedData()` is something that is _not_ specific to the loading user (i.e. every can/should see the same data) and can be determined purely from the path to the page, then yes `getStaticProps` is a good replacement for `useState`.

Now the data will be available immediately rather than after a second network request.

If the data cannot ever be stale or must be customized to every user, then it doesn't work as well.

If state-while-refresh is ok, you can mix the two. Use `getStaticProps` to have _something_ rendered at first, and do a network request to fetch the latest up-to-date data and overwrite the initial props.

If data must be customized for every user, you could check out alternatives like `getInitialProps` or `getServerSideProps` for per-request loading, that happens server-side so the first render still has data available
